proteus/LinearAlgebraTools.py

```diff
@@ -272,6 +307,126 @@ def mult(self, A, x, y):
             self.ghosted_csr_mat.matvec(xlf.getArray(),ylf.getArray())
         y.setArray(self.yGhosted.getArray())
 
+class MatrixShell:
+    """ A shell class for a matrix. """
+    def __init__(self,A):
+        """
+        Specifies a basic matrix shell.
+
+        Parameters
+        ----------
+        A : matrix
+            A petsc4py matrix object
+        """
+        self.A = A
+    def create(self,A):
+        pass
+    def mult(self,A,x,y):
+        """
+        Multiply the matrix and x.
+
+        Parameters
+        ----------
+        A : matrix
+            Dummy place holder for PETSc compatibility
+        x : vector
+
+        Returns
+        -------
+        y : vector
+        """
+        self.A.mult(x,y)
+
+class MatrixInvShell:
+    """ A PETSc shell class for a inverse operator. """
+    def __init__(self, A):
+        """
+        Initializes operators and solvers for inverse operator.
+
+        Parameters
+        ----------
+        A : PETSc matrix
+            This is the matrix object used to construct the inverse.
+        """
+        self.A = A
+        self.ksp = p4pyPETSc.KSP().create()
+        self.ksp.setOperators(self.A,self.A)
+        self.ksp.setType('preonly')
+        self.ksp.pc.setType('lu')
+        self.ksp.setUp()
+    def create(self,A):
+        pass
+    def apply(self,A,x,y):
+        """
+        Apply the inverse pressure mass matrix.
+
+        Parameters
+        ----------
+        A : matrix
+            Dummy place holder for PETSc compatibility
+        x : vector
+
+        Returns
+        -------
+        y : vector
+        """
+        self.ksp.solve(x,y)
+
+class PCDInv_shell:
```

Comment:
probably want to delete "pressure mass" in the docstring, since this is a generic inverse matrix shell